deep_quoridor/src/quoridor.py

```diff
@@ -0,0 +1,293 @@
+
+        try:
+            self._grid[*old_position * 2] = Board.FREE
+            self._grid[*new_position * 2] = player
+            self._player_positions[player] = new_position
+        except IndexError:
+            return Board.OUT_OF_BOUNDS
+
+    def get_cell(self, position: Position) -> int:
+        """
+        Get the value of a "player cell" in the grid.
+        """
+        position = np.asarray(position, dtype=np.uint8)
+        try:
+            return self._grid[*position * 2]
+        except IndexError:
+            return Board.OUT_OF_BOUNDS
+
+    def add_wall(self, player: int, position: Position, orientation: WallOrientation):
+        """
+        Mark the grid cells corresponding to the wall as occupied.
+        """
+        position = np.asarray(position, dtype=np.uint8)
+
+        if self._walls_remaining[player] < 1:
+            raise RuntimeError("Cannot place wall, no walls remaining for player {player}")
+        try:
+            self._grid[self._get_wall_slices(position, orientation)] = 10 + player
+        except IndexError:
+            return Board.OUT_OF_BOUNDS
+
+        self._walls_remaining[player] -= 1
+
+    def can_place_wall(self, position: Position, orientation: WallOrientation) -> bool:
+        """
+        Returns True if the wall can be placed at the given position and orientation.
+
+        Checks that the player has walls remaining, the wall is within the bounds of the board, and doesn't overlap with other walls.
+        """
+        position = np.asarray(position, dtype=np.uint8)
+        try:
+            return (self._grid[self._get_wall_slices(position, orientation)] == Board.FREE).all()
+        except IndexError:
+            return False
+
+    def is_wall_between(self, position1: Position, position2: Position) -> bool:
+        """
+        Check if there is a wall between two positions.
+        """
+        position1 = np.asarray(position1, dtype=np.uint8)
+        position2 = np.asarray(position2, dtype=np.uint8)
+        wall_position = position1 + position2
+        return self._grid[*wall_position] == Board.PLAYER1_WALL
+
+    def _get_wall_slices(self, position: Position, orientation: WallOrientation) -> tuple[slice, slice]:
+        """
+        Get a tuple of slices that correspond to the wall's cells in the grid.
+        """
+        if orientation == WallOrientation.VERTICAL:
+            return (slice(position[0] * 2, position[0] * 2 + 3), position[1] * 2 + 1)
+        elif orientation == WallOrientation.HORIZONTAL:
+            return (position[0] * 2 + 1, slice(position[1] * 2, position[1] * 2 + 3))
+        else:
+            raise ValueError("Invalid orientation: {orientation}")
+
+    def __str__(self):
+        """
+        Return a pretty-printed string representation of the grid.
+        """
+        display_grid = np.full(self._grid.shape, " ", dtype=str)
+        display_grid[::2, ::2] = "."
+        display_grid[self._grid == Board.PLAYER1_WALL] = "w"
```

Comment:
Is there a reason we are keeping track of which player placed which wall where? I don't think there is anywhere we will use that info and it complicates the data model.

Reply:
It's for display. I think (especially when debugging an agent) it will be useful to be able to color the walls differently based on who placed them.